lib/Device.js

```diff
@@ -237,19 +238,26 @@ class Device extends Entity {
     // Returns:                 Nothing
     assign(options) {
         const deviceGroup = new DeviceGroup(options);
-        this.getEntity().
+        let devices = options.deviceIdentifier.map(deviceId => new Device().initByIdentifier(deviceId));
+        this._helper.makeConcurrentOperations(devices, device => device.getEntity()).
```

Comment:
How does this scale?  It looks like `impt device unassign | assign` has underlying support in [imp-central-api](https://github.com/electricimp/imp-central-api/) for doing multiple device operations...  However, `assign` is using this `makeConcurrentOperations` to run some preflight checks.  It feels like there are some opportunities for [rate limiting](https://apidoc.electricimp.com/#section/Rate-Limiting) and other problems as we scale up the number of devices.  In poking at impCentral, it appears that my accounts `X-RateLimit-Limit` = **50**, making the total number that you could do in a single command to be less than that...

I don't need to support hundreds of thousands of imps with these commands (at least not yet!), but 10's or 100's doesn't feel out of the question and I'm concerned that as implemented this will not work.  Number of imps could make this challenging, but did we do any scale testing here?

One thought for a workaround would be a CLI flag to skip the preflight checks?  Perhaps `-q / confirmed` should be have its functionality tweaked to skip all the requests except for `assignToDG` ?

Reply:
X-RateLimit error is processed and workarounded by impt. The requests are executed in portions (not all at once in parallel) and repeated with pauses till accepted by the server. It was tested with hundreds of entities (not with devices but with deployments - it does not matter as the processing is common).
So, it should not be a problem from workability point of view, as we hope. Please let us know, if/when you run it for hundreds of devices.
 
The only possible concern is performance. "Preflight" checks add requests. But it is a common solution for many different entities. It allows to specify an entity not by Id only. Eg. you can use AgentId, MAC address or even Name instead of deviceId.
An update here will be a significant update of the whole design. We would consider that only if the performance is totally unacceptable.